### PR TITLE
feat: integrated app upgrade from App Hub and update UI

### PR DESCRIPTION
**Pull request by mediremi**
Closes https://jira.dhis2.org/browse/DHIS2-10501 and https://jira.dhis2.org/browse/DHIS2-122

### CODE REVIEW

File: src/components/CustomAppDetails/index.js
Comment (amcgee) [inline src/components/CustomAppDetails/index.js:194]: We support dynamic query `id`, `params`, and `data` properties with a static query using query variables, so this can be rewritten like so:

```js
const query = {
    app: {
        resource: `appHub/v1/apps`,
        id: ({ appHubId }) => appHubId,
    },
    installedApps: {
        resource: 'apps',
    },
}
const CustomAppDetails = ({ match }) => {
    const { loading, error, data, refetch } = useDataQuery(query, { variables: { appHubId: match.params.appHubId } })
```

File: src/components/CoreApps/index.js
Comment (amcgee) [inline src/components/CoreApps/index.js:67]: I think we should probably enforce *only* matching by app hub ID.  That will require us to set the app hub id for all core apps before releasing 2.36 which is a very small, low-risk change but also needs all apps on the app hub (which we should probably do in any case).

What do you think?  Leaving the "soft fallback" for bundled apps does give us more time to get the infrastructure set up...

If we keep this "fallback" it should only be supported for apps where `isBundled === true`, I think.

Comment (mediremi) [reply src/components/CoreApps/index.js:67]: Yep I agree let's do only matching by App Hub ID. I don't see why we wouldn't be able to set the App Hub ID for all core apps before 2.36.

**Review by amcgee — CHANGES_REQUESTED** [review]
@mediremi the UI changes here look really good.

However, there is unfortunately no way we can merge it as-is, since it **removes the ability to install apps from the App Hub**.  I know you're working on the App Hub components which will eventually be replicated here (some already are) but we can't lose features in the meantime (especially as this will be the version included in 

Additionally, in the new app UI there is **no way to uninstall custom apps**, which is also a feature regression and needs to be fixed.

I think we should probably keep this restricted in scope to just adding the **check for updates** logic in 2.36 as the hard freeze is upon us today...

File: src/components/CoreApps/index.js
Comment (amcgee) [inline src/components/CoreApps/index.js:46]: This needs to use `short_name` NOT `name`.  `short_name` is guaranteed to be unique and core app short names are protected.  Any app can use the same `name` as a core app without actually overriding that core app.

File: src/components/CoreApps/index.js
Comment (amcgee) [inline src/components/CoreApps/index.js:25]: These should be, again, short names rather than names.  They should then be prefixed with `dhis-web` to get the namespace of that application (i.e. `dhis-web-app-management`, `dhis-web-capture`, etc - look in the URL when visiting an app to see this).

This is also missing several bundled / core applications, notably the STRUTS apps - the full list is [here](https://github.com/dhis2/dhis2-core/blob/master/dhis-2/dhis-api/src/main/java/org/hisp/dhis/appmanager/AppManager.java#L49-L81)

Finally - when this switches to using `short_name` we should be fetching the translations of bundled applications from the Core (POST /api/translations or displayName from GET /dhis-web-commons/menu/getModules.action), rather than always showing them in English) - this could be a future enhancement if it's too much work for now as no other apps are currently translated, but would be a nice-to-have.

File: src/components/CoreApps/index.js
Comment (amcgee) [inline src/components/CoreApps/index.js:54]: (see above) short name is NOT the same as name!

File: src/components/CoreApps/index.js
Comment (amcgee) [inline src/components/CoreApps/index.js:58]: This might be OK for the first pass but is not ideal - this really should show the icon for the bundled application, which should be available at `/icons/dhis-web-${short_name}.png` and is definitely included in the getModules.action response (ugly I know, but it's what we have)

File: src/components/CustomAppDetails/Versions.js
Comment (amcgee) [inline src/components/CustomAppDetails/Versions.js:101]: This doesn't make a lot of sense here, as only versions compatible with the current server version should be available.  It might be nice to show "disabled" versions which are incompatible, but the select box doesn't make much sense.

Also, showing the current and two previous versions is incorrect.  I will comment separately there.

File: src/components/CustomAppDetails/Versions.js
Comment (amcgee) [inline src/components/CustomAppDetails/Versions.js:216]: This has a few different issues:

1. As mentioned above, having a select doesn't really make sense at all when the point is to install into the current instance so this should probably be removed
2. This actually gets the current server version which **is almost never actually the latest version**, so showing the current version and two previous releases is incorrect.
3. When referencing the version of the current server we should use [`serverVersion`](https://github.com/dhis2/app-runtime/blob/master/services/config/src/types.ts#L1-L6) from the `config` object instead of `apiVersion` (they currently match but this might not always be true)

File: src/components/CustomAppDetails/Versions.js
Comment (amcgee) [inline src/components/CustomAppDetails/Versions.js:1]: This file has many different components and mixes a lot of business logic with presentation logic, from a code style and readability perspective I would like to see them separated out.

File: src/components/CustomAppDetails/Versions.js
Comment (amcgee) [inline src/components/CustomAppDetails/Versions.js:84]: I don't know that we should show this here... we don't want to promote development and canary builds to users who don't really know what they mean.  I think since we only have stable builds at the moment we should omit this set of checkboxes entirely.

File: src/components/AppList/index.js
Comment (amcgee) [inline src/components/AppList/index.js:30]: (minor, could be fixed later) - we do want to show details about installed apps even if they don't have an appHub ID - version, developer, etc (things from the manifest) still make sense here.

**Review by amcgee — CHANGES_REQUESTED** [review]
Left some specific comments here as well - whether we try to address these today or revert to an updates-only commit (or just fix the update version parsing logic in the self-updater) is up to you @mediremi - there may be some more bugs or usability issues to discover, but these are the ones that jumped out at me.

**Comment by amcgee** [comment]
A few minor general code style comments:
1. We should avoid putting code in `index.js` files - we reserve these for re-exports.
2. Files should generally have only one logical function (i.e. especially if it is a React component) and should be named the same as that function or component.  Occasionally a module which exports several small helper functions or components could make sense - this isn't a hard and fast rule - but it helps to grep the code and understand what a file does more quickly.
3. I personally prefer named to default exports (unambiguous es-module / commonjs interchange semantics and sometimes better editor code completion, plus explicitly unambiguous import naming).  This isn't an established standard, though, so fine if you prefer defaults.

Again, really great work on this and the UI looks (and functions) really well!  Please excuse my late-night nitpicking :-)

**Comment by mediremi** [comment]
> However, there is unfortunately no way we can merge it as-is, since it removes the ability to install apps from the App Hub. I know you're working on the App Hub components which will eventually be replicated here (some already are) but we can't lose features in the meantime (especially as this will be the version included in

> I think we should probably keep this restricted in scope to just adding the check for updates logic in 2.36 as the hard freeze is upon us today

:+1: I'll create a PR based off `master` that just adds the check for update logic and I'll revisit this PR later

File: src/get-latest-version.js
Comment (mediremi) [inline src/get-latest-version.js:3]: Should we only return the latest version from the stable channel? Or take the channel as a parameter so that we show users only the latest version for the channel their app is from?

File: src/pages/CoreApps/CoreApps.js
Comment (mediremi) [inline src/pages/CoreApps/CoreApps.js:17]: Instead of fetching every App Hub app in order to see if any installed apps have pending updates, maybe we should have a `v2` API that given a list of app IDs returns data about those apps.

File: src/pages/AppHub/AppHub.js
Comment (mediremi) [inline src/pages/AppHub/AppHub.js:23]: The DHIS version should be passed as a parameter so that only compatible apps are shown.

File: src/core-apps.js
Comment (mediremi) [inline src/core-apps.js:1]: We should also hardcode the App Hub ID of each app so that matching of core apps to their App Hub version can be done without comparing names/organisations.

File: src/pages/CoreApps/CoreApps.js
Comment (mediremi) [inline src/pages/CoreApps/CoreApps.js:80]: We need to verify that the `minDhisVersion` and `maxDhisVersion` fields of the versions are suitable for the instance version.

File: src/pages/CustomApps/CustomApps.js
Comment (mediremi) [inline src/pages/CustomApps/CustomApps.js:55]: DHIS version checking needs to happen here too.

**Comment by amcgee** [comment]
@mediremi converted this to a draft for now

File: src/pages/CoreApps/CoreApps.js
Comment (amcgee) [reply src/pages/CoreApps/CoreApps.js:17]: Yeah, definitely agreed.  We could just request each app independently for now, and maybe implement a graphql API in the future to support batching.

Comment (amcgee) [reply src/pages/CoreApps/CoreApps.js:17]: It would be nice to get a "count" of updates available (for each core apps and installed apps) though, to show a badge in the sidebar or similar.  That would also allow us to more easily implement an "update all" button

File: src/get-latest-version.js
Comment (amcgee) [reply src/get-latest-version.js:3]: Channels from the perspective of the DHIS2 instance and App Management App needs some more thought I think...

File: src/pages/CoreApps/CoreApps.js
Comment (amcgee) [reply src/pages/CoreApps/CoreApps.js:80]: Or maybe pass the current DHIS2 version?

File: src/pages/CoreApps/CoreApps.js
Comment (Birkbjo) [inline src/pages/CoreApps/CoreApps.js:59]: Would it be easier if we were to add a `core`-property to the API response from the App Hub? Or would we need to do this matching regardless?

File: src/pages/AppHub/AppHub.js
Comment (Birkbjo) [inline src/pages/AppHub/AppHub.js:76]: Any reason for this not being translated?

Would also be great to have a more "forgiving"/specialized message on the index-page (Preinstalled Core apps). Might not be worth the extra-props/complexity of reusing this component though, as Core apps should hopefully be on the App Hub before this is released.

File: src/pages/CoreApps/CoreApps.js
Comment (mediremi) [reply src/pages/CoreApps/CoreApps.js:59]: Yep adding a `core` property to the API response would eliminate the need for the `src/core-apps.js` file and the `coreApps.find(({ name }) => name === app.name)` matching.

File: src/pages/AppHub/AppHub.js
Comment (mediremi) [reply src/pages/AppHub/AppHub.js]: Nope - good catch :+1: 

The `AppsList` component already has props for custom messages/labels to use for available updates, headings and search input label so having an additional prop for the 'no apps match your criteria' message should be fine.

**Review by Birkbjo — APPROVED** [review]

**Review by varl — APPROVED** [review]

**Comment by amcgee** [comment]
Should we get this merged today or this week?  Still waiting on the CI integration?

**Comment by amcgee** [comment]
Also, have we tested this well against all supported DHIS2 versions?  (Should be 2.35+ I think)

**Comment by mediremi** [comment]
> Should we get this merged today or this week? Still waiting on the CI integration?
It'd be nice to merge this after setting up CI integration but since we probably want this merged before tomorrow's presentation we could merge now :man_shrugging: 

> Also, have we tested this well against all supported DHIS2 versions? (Should be 2.35+ I think)
Installing, uninstalling as well as the filtering of core/custom apps has been tested on 2.35, 2.36 and 2.37 :heavy_check_mark:

**Comment by varl** [comment]
> Should we get this merged today or this week? Still waiting on the CI integration?

CI for App Hub publish ? The rest should be set up.

**Comment by amcgee** [comment]
@varl yeah CI for publish

**Comment by amcgee** [comment]
I think I'd advocate for merging this and publishing to App Hub manually (making sure we set a decently incremented version number) for today/tomorrow presentations

**Comment by amcgee** [comment]
Would obviously be nice to have a publish workflow in place though

**Pull request merged**
